ld.lld error: undefined symbol: reallocarray

### Zig Version

0.12.0-dev.306+a0968be83

### Steps to Reproduce and Observed Behavior

```
]$ cat main.c
#define _DEFAULT_SOURCE 1
#define _GNU_SOURCE 1

#include <stdlib.h>

int main () {
  char* buffer = malloc(12);
  buffer = reallocarray(buffer, 1, 13);
  free(buffer);
}
```

This may or may not link, depending on the Zig version, etc:

```
$ ./zig-11 version
0.11.0

$ ./zig-12 version
0.12.0-dev.306+a0968be83

$ ./zig-12 cc -lc main.c
ld.lld: error: undefined symbol: reallocarray
>>> referenced by main.c:8
>>>               /home/acarrico/.cache/zig/o/3bc394cabb7b4314ac3135fa11f380c2/main.o:(main)

$ ./zig-11 cc -lc main.c
ld.lld: error: undefined symbol: reallocarray
>>> referenced by main.c:8
>>>               /home/acarrico/.cache/zig/o/25067a08a7699e4e9c5a4ac69ebc8c85/main.o:(main)

$ ./zig-12 cc -lc main.c --target=x86_64-linux-gnu
ld.lld: error: undefined symbol: reallocarray
>>> referenced by main.c:8
>>>               /home/acarrico/.cache/zig/o/cbc6b9a81ea2a088144021ffd58635b5/main.o:(main)

$ ./zig-12 cc -lc main.c --target=x86_64-linux-musl

$ ./zig-11 cc -lc main.c --target=x86_64-linux-gnu
ld.lld: error: undefined symbol: reallocarray
>>> referenced by main.c:8
>>>               /home/acarrico/.cache/zig/o/e42380168d6a5487bca5dd7cac83b81f/main.o:(main)

$ ./zig-11 cc -lc main.c --target=x86_64-linux-musl

$ ln -s ~/no-backup/zig-linux-x86_64-0.11.0-dev.1366+c70a3d902/zig zig-11-dev

$ ./zig-11-dev cc -lc main.c

$ ./zig-11-dev cc -lc main.c --target=x86_64-linux-musl

$ ./zig-11-dev cc -lc main.c --target=x86_64-linux-gnu
ld.lld: error: undefined symbol: reallocarray
>>> referenced by main.c:8
>>>               /home/acarrico/.cache/zig/o/d37587f0b76e20739639ab1a898c73df/main.o:(main)

$ nix-shell -p zig

[nix-shell]$ zig version
0.10.1

[nix-shell]$ zig cc -lc main.c

[nix-shell]$ zig cc -lc main.c --target=x86_64-linux-gnu
ld.lld: error: undefined symbol: reallocarray
>>> referenced by main.c:8
>>>               /home/acarrico/.cache/zig/o/daa499687d5d9e4c86360cc7a0fd8e66/main.o:(main)

[nix-shell]$ zig cc -lc main.c --target=x86_64-linux-musl

[nix-shell]$
```


### Expected Behavior

reallocarray should be available with ```-lc```. It seems particularly strange that it is missing under gnu, since I believe it has been there longer than in musl. It seems to be available in older versions without specifying a target (perhaps this is linking the system's libc?).

My suggested fix in #17610 turns the link error into a compile time error:
```
$ zig  cc main.c -target x86-linux-gnu
main.c:8:12: error: call to undeclared function 'reallocarray'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    8 |   buffer = reallocarray(buffer, 1, 13);
      |            ^
main.c:8:10: error: incompatible integer to pointer conversion assigning to 'char *' from 'int' [-Wint-conversion]
    8 |   buffer = reallocarray(buffer, 1, 13);
      |          ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
```

The fix for your code is to use an explicit, later version of glibc:
```
$ zig cc main.c -target x86-linux-gnu.2.38
```

--- a/lib/libc/include/generic-glibc/malloc.h
+++ b/lib/libc/include/generic-glibc/malloc.h
@@ -51,6 +51,11 @@ __THROW __attribute_malloc__ __attribute_alloc_size__ ((1, 2)) __wur;
 extern void *realloc (void *__ptr, size_t __size)
 __THROW __attribute_warn_unused_result__ __attribute_alloc_size__ ((2));
 
+/*
+ * reallocarray introduced in glibc 2.26
+ * https://sourceware.org/git/?p=glibc.git;a=commit;h=2e0bbbfbf95fc9e22692e93658a6fbdd2d4554da
+ */
+#if (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 26) || __GLIBC__ > 2
 /* Re-allocate the previously allocated block in PTR, making the new
    block large enough for NMEMB elements of SIZE bytes each.  */
 /* __attribute_malloc__ is not used, because if reallocarray returns
@@ -59,6 +64,7 @@ __THROW __attribute_warn_unused_result__ __attribute_alloc_size__ ((2));
 extern void *reallocarray (void *__ptr, size_t __nmemb, size_t __size)
   __THROW __attribute_warn_unused_result__ __attribute_alloc_size__ ((2, 3))
   __attr_dealloc_free;
+#endif /* glibc v2.26 and later */
 
 /* Free a block allocated by `malloc', `realloc' or `calloc'.  */
 extern void free (void *__ptr) __THROW;

--- a/lib/libc/include/generic-glibc/stdlib.h
+++ b/lib/libc/include/generic-glibc/stdlib.h
@@ -686,6 +686,11 @@ extern void *realloc (void *__ptr, size_t __size)
 /* Free a block allocated by `malloc', `realloc' or `calloc'.  */
 extern void free (void *__ptr) __THROW;
 
+/*
+ * reallocarray introduced in glibc 2.26
+ * https://sourceware.org/git/?p=glibc.git;a=commit;h=2e0bbbfbf95fc9e22692e93658a6fbdd2d4554da
+ */
+#if (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 26) || __GLIBC__ > 2
 #ifdef __USE_MISC
 /* Re-allocate the previously allocated block in PTR, making the new
    block large enough for NMEMB elements of SIZE bytes each.  */
@@ -701,6 +706,7 @@ extern void *reallocarray (void *__ptr, size_t __nmemb, size_t __size)
 extern void *reallocarray (void *__ptr, size_t __nmemb, size_t __size)
      __THROW __attr_dealloc (reallocarray, 1);
 #endif
+#endif /* glibc v2.26 and later */
 
 #ifdef __USE_MISC
 # include <alloca.h>

